### PR TITLE
Reverse vector mode

### PR DESCRIPTION
**Pull request by tgymnich**

### CODE REVIEW

File: enzyme/Enzyme/AdjointGenerator.h [hunk 1]
```diff
@@ -1735,6 +1786,14 @@ class AdjointGenerator
     ((DiffeGradientUtils *)gutils)->applyChainRule(Builder, rule, args...);
   }
 
+  /// Unwraps an collection of constant vector derivatives from their internal
+  /// representations and applies a function f to each element.
+  template <typename Func>
```
Comment (wsmoses) [inline enzyme/Enzyme/AdjointGenerator.h:1791]: Covered already?

File: enzyme/Enzyme/GradientUtils.h [hunk 2]
```diff
@@ -2098,7 +2098,7 @@ class DiffeGradientUtils : public GradientUtils {
         idx2.push_back(v);
         auto selects = addToDiffe(
             val, BuilderM.CreateExtractValue(dif, ArrayRef<unsigned>(i)),
-            BuilderM, nullptr, idx2);
+            BuilderM, addingType, idx2);
```
Comment (wsmoses) [inline enzyme/Enzyme/GradientUtils.h:2101]: Save for later? This seems like the wrong way to handle (and worth explicitly testing)

File: enzyme/Enzyme/GradientUtils.h [hunk 3]
```diff
@@ -2117,7 +2117,7 @@ class DiffeGradientUtils : public GradientUtils {
...
```
Comment (wsmoses) [inline enzyme/Enzyme/GradientUtils.h:2120]: Same comment

Comment (wsmoses) [reply enzyme/Enzyme/GradientUtils.h:2120]: E.g. I would've expected addingType to also be an array and needing to extract the inner type as the inner addingType

Comment (wsmoses) [reply enzyme/Enzyme/GradientUtils.h:2120]: That said usually the result here comes directly from TA and that probably gives a scalar -- but this is likely why this was nullptr.

**Review by wsmoses — APPROVED** [review]
LGTM, but make sure Julia passes before merging.

File: enzyme/Enzyme/Enzyme.cpp [hunk 4]
```diff
@@ -970,11 +970,19 @@ class Enzyme : public ModulePass {
...
+      else if (cast<Function>(fn)->getReturnType()->isFPOrFPVectorTy()) {
+        Constant *seed =
+            ConstantFP::get(cast<Function>(fn)->getReturnType(), 1.0);
+        if (width == 1) {
+          args.push_back(seed);
+        } else {
+          ArrayType *arrayType =
+              ArrayType::get(cast<Function>(fn)->getReturnType(), width);
+          args.push_back(
+              ConstantArray::get(arrayType, SmallVector(width, seed)));
```
[the remark this answers is not in the document]
Comment (tgymnich) [reply enzyme/Enzyme/Enzyme.cpp]: done

[the remark this answers is not in the document]
Comment (tgymnich) [reply enzyme/Enzyme/Enzyme.cpp]: done

**Pull request merged**